File: backend/app/models/concept.py
"""Concept node models.

A concept is a source-independent evergreen idea ("compounding", "loss aversion")
mined from every card and deduplicated across the whole library. Mirrors the
artifact/catalog subsystem at every layer.
"""

from __future__ import annotations

import uuid
from datetime import datetime, timezone
from typing import Optional

from pydantic import BaseModel, Field


def _new_id() -> str:
    return "c_" + uuid.uuid4().hex[:8]


class Concept(BaseModel):
    """One concept name, as emitted by structuring (pre-aggregation)."""

    name: str


class ConceptEntry(BaseModel):
    """An aggregated, deduplicated concept: one idea, many source cards."""

    id: str = Field(default_factory=_new_id)
    name: str
    source_card_ids: list[str] = Field(default_factory=list)
    definition: Optional[str] = None
    created_at: str = Field(
        default_factory=lambda: datetime.now(timezone.utc).isoformat()
    )

File: app/lib/domain/models/concept.dart
/// Concept node: a source-independent evergreen idea mined across all cards.
/// Mirrors the backend concept subsystem (models/concept.py).
library;

class ConceptEntry {
  const ConceptEntry({
    required this.id,
    required this.name,
    this.sourceCardIds = const [],
    this.definition,
  });

  final String id;
  final String name;
  final List<String> sourceCardIds;

  /// On-demand LLM definition; null until "Define" is tapped.
  final String? definition;

  factory ConceptEntry.fromJson(Map<String, dynamic> json) => ConceptEntry(
        id: (json['id'] as String?) ?? '',
        name: (json['name'] as String?) ?? '',
        sourceCardIds: (json['source_card_ids'] as List?)
                ?.map((e) => e.toString())
                .toList() ??
            const [],
        definition: json['definition'] as String?,
      );
}

class ConceptDetail {
  const ConceptDetail({required this.entry, this.related = const []});

  final ConceptEntry entry;
  final List<ConceptEntry> related;

  factory ConceptDetail.fromJson(Map<String, dynamic> json) => ConceptDetail(
        entry: ConceptEntry.fromJson(
            (json['entry'] as Map<String, dynamic>?) ?? {}),
        related: (json['related'] as List?)
                ?.whereType<Map<String, dynamic>>()
                .map(ConceptEntry.fromJson)
                .toList() ??
            const [],
      );
}
